Move graph and font

stefan-b-jakobsson opened the pull request:
NOTICE!!! This PR needs to be reviewed and tested before merge!

This PR moves the graph and fonts functions from ROM bank 0 to a separate ROM bank (8). The test demo code is moved from ROM bank 8 to ROM bank 9. The default framebuffer driver remains in ROM bank 0.

The move is made to free up space in ROM bank 0, which is needed for future development of the Kernal, as described by @mist64 in issue #316.  Currently there are literally only a handful of bytes free. This PR frees up about 5.5 kB in ROM bank 0. 

The ROM bank 0 vectors for graph and font functions are still there not to break any existing code. These bank 0 functions are, however, only wrappers that call code in bank 8.

Multiple bank switching has a negative effect on performance, and there is some performance to gain from calling the bank 8 functions directly instead of through the bank 0 vectors. I'm not too worried about performance though, as I don't think those who seek performance will be using the built-in graph and fonts library.

At the beginning of ROM bank 8, there's a jump table for all externally exposed functions. This jump table is available in the include file inc/graphics.inc.

The graph and font functions need to call a few Kernal functions and use a few Kernal RAM variables that have no fixed address. These symbols and their memory addresses are determined during the link stage by the utility "findsymbols" that I created. This utility retrieves memory addresses of symbols from a .sym file. The utility is a simple bash script that invokes awk. Thanks to @ZeroByteOrg for suggesting this solution.

In the current master branch there is a bug that affects the primary foreground color setting for graph functions. The primary color is held in RAM variable col1, but this memory address is also used by the Kernal for a different purpose (activity_led_state). This bug has been fixed in the PR.

The new code runs the built-in tests: TEST 0, TEST 1 and TEST 2 without any apparent problems. It is also possible to start GEOS, but I have no GEOS system disk to actually test the system with.

ZeroByteOrg commented:
I hope this PR is approved, as I would also like to make a PR for updating LOAD to use MACPTR, but cannot until the image has sufficient space to work with. (my work-in-progress just deleted the font to make room while developing and testing the change)

stefan-b-jakobsson commented:
I hope it can be accepted even though it's a huge push of almost 1,900 line of code.
It's not as bad as it might first look though. Most of the code is just moved with a little glue in between.
As there initially was a lot of trial and error, I wasn't skilled enough to document my work in small incremental commits.

stefan-b-jakobsson commented:
After some bug fixes, moving graph and font from bank 0 to bank 8 is working both in the emulator and on hardware.

To gain some performance, and to free up some more space in bank 0, I committed code that moves the default frame buffer driver from bank 0 to bank 8.

Keeping graph and font in bank 8 and the framebuffer driver in bank 0 caused the Kernal to do a lot of bank switching which had a significant impact on speed. According to a simple test I did, there was a 50 to 60 % increase of execution time.

After moving the default framebuffer driver to bank 8, I see performance about 3 percent less than the current master branch.

After this change, there is about 6.3 kB free in ROM bank 0.

This new code also needs to be tested on hardware before considered for merge.

irmen commented:
The PR won't compile for me (building on Linux):

caused by 8636f16b02a8c204e379d3b4a76f84c64aec088c where the ``kernal.sym`` filename is erroneously called ``Kernal.sym`` twice in the makefile's grap.bin target step.  This fails on case-sensitive filesystems.

irmen commented on Makefile:
```diff
@@ -423,7 +433,12 @@ $(BUILD_DIR)/charset.bin: $(CHARSET_OBJS) $(CHARSET_DEPS) $(CFG_DIR)/charset-$(M
+# Bank 8 : Graphics
+$(BUILD_DIR)/graph.bin: $(GRAPH_OBJS) $(KERNAL_DEPS) $(CFG_DIR)/graph.cfg
+	@mkdir -p $$(dirname $@) 
+	$(LD) -C $(CFG_DIR)/graph.cfg $(GRAPH_OBJS) -o $@ -m $(BUILD_DIR)/graph.map -Ln $(BUILD_DIR)/graph.sym `${BUILD_DIR}/../../findsymbols ${BUILD_DIR}/Kernal.sym col1 col2 col_bg ptr_fg` `${BUILD_DIR}/../../findsymbols ${BUILD_DIR}/Kernal.sym -p k_ kbdbuf_get sprite_set_image sprite_set_position`
```
should be `kernal.sym` (lowercase k) twice

stefan-b-jakobsson replied:
Pushed a fix for that. My machine has a case insensitive file system and built fine before the fix as well.

mist64 commented:
I'm looking at this right now. (Sorry for the delay!)

mist64 commented on kernal/graph/graph.s:
```diff
@@ -1,606 +1,68 @@
+.export GRAPH_set_colors
+.export GRAPH_set_window
 
+.export col1,col2,col_bg
```
instead of declaring "graph" variables in "kernal" source and exporting them, wouldn't it be easier to declare a `GRAPHVAR` segment in cfg/x16.cfginc (growing from the top down into the 6 byte "reserved for KERNAL growth" hole) and declaring these in the "graph" source?

This would be consistent with what we do with ROM banks. Other ROM banks have their section of variable space, named after the ROM, declared in cfg/x16.cfginc and filled in the respective source.

stefan-b-jakobsson replied:
It would be cleaner if these variables were defined in the graph bank.
I think I left them there not to affect any following symbol addresses guarded by asserts.
If it's important we can move them, but it might require some work...

mist64 reviewed (commented):
Sorry I'm quite nitpicky here.

I do appreciate very clean and pretty fix to the ROM's biggest problem! :)

mist64 commented on kernal/fonts/fonts.s:
```diff
@@ -1,62 +1,62 @@
-
-.include "font_internal.inc"
-
-.assert curIndexTable = k_curIndexTable, error, "update font_internal.inc!"
```
(That said, if we leave this here, it probably should have a comment saying what I explained above, so it's clear why this is needed in the first place...)

stefan-b-jakobsson replied:
I will restore the asserts.

stefan-b-jakobsson commented on cfg/x16.cfginc:
```diff
@@ -38,5 +39,5 @@ BVARS:    start = $03D4, size = $002C; # BASIC
 /* KERNAL/DOS bank #0 vars */
 KEYMAP:   start = $A000, size = $0800; # the current keyboard mapping table
-KVARSB0:  start = $A800, size = $0800; # there is a lot of space free here
+KVARSB0:  start = $A800, size = $0700; # there is a lot of space free here
```
I shrunk KVARSB0 used in bank 0 by by 0x100 bytes (0xa800-0xaeff).
This space (0xaf00-afff) is used by the graphics/fonts bank (memory segment GRVARSB0), especially the buffers declared in graphics/graph/console.s

stefan-b-jakobsson commented on kernal/drivers/x16/smc.s:
```diff
@@ -3,10 +3,9 @@
 
-.segment "KVAR"
-
+.segment "KVAR1a"
```
KVAR1a is still used in kernal/drivers/x16/smc.s
It was activity_led_state defined there that had the .res 0
I put that in KVAR1a not to affect any other symbol addresses.

stefan-b-jakobsson commented on kernal/fonts/fonts.s:
```diff
@@ -1,62 +1,62 @@
 baselineOffset:	.res 1
 curSetWidth:	.res 2
-curHeight:	.res 1
+curHeight:	    .res 1
```
I guess the problem here is that my editor inserted spaces instead of tabs. But isn't the tab width 4?

stefan-b-jakobsson commented:
Thanks @mist64 for reviewing the code. Please let me know what you think we should do with the items I could not easily fix above.

stefan-b-jakobsson commented on basic/x16additions.s:
```diff
@@ -626,7 +626,7 @@ test:
 @copy:
 	sei
-	lda #8
+	lda #9
```
Fixed and pushed.

stefan-b-jakobsson commented:
I did mange to do some clean-up of the Kernal vars in commit 8351e48 above:

- Variables col1, col2 and col_bg are now defined in the GRAPH bank code
- I removed the memory segment KVAR1a (1 byte) and moved that to KVAR, which was effectively reduced by 2 bytes
- The new memory segment GRAPHVAR holds col1, col2 and col_bg
- I added some symbols to kernal/fonts/font_internal.inc used by the GRAPH routines
- I changed graph/fonts/font_internal.inc to just include the contents of kernal/fonts/font_internal.inc
- Assertions were updated

stefan-b-jakobsson commented on cfg/x16.cfginc:
```diff
@@ -20,8 +20,9 @@ ZPMATH:   start = $00A9, size = $002B; # MATH
-KVAR:     start = $0200, size = $00BD; # KERNAL
-/*        start = $02BD, size = $0007; # reserved for KERNAL growth */
+KVAR:     start = $0200, size = $00BB; # KERNAL
+/*GRAPHVAR: start = $02BB, size = $0003; # GRAPH BANK VARS*/
```
Done
